packages/core/src/model.types.ts

```diff
@@ -38,9 +40,18 @@ export interface Model<
   actions: Prop<TModelCreators, 'actions'>;
   reset: () => AssignAction<TContext, any>;
   createMachine: (
-    config: MachineConfig<TContext, any, TEvent, TAction>,
-    implementations?: Partial<MachineOptions<TContext, TEvent, TAction>>
-  ) => StateMachine<TContext, any, TEvent>;
+    config: MachineConfig<TContext, any, TEvent, TExtra, TAction>,
+    implementations?: Partial<MachineOptions<TContext, TEvent, TExtra, TAction>>
+  ) => StateMachine<
+    TContext,
+    any,
+    TEvent,
+    { value: any; context: TContext },
+    TExtra
+  >;
+  withInput: <TInput extends {}>(
+    input: TInput
+  ) => Model<TContext, TEvent, TAction, TModelCreators, { input: TInput }>;
```

Comment:
I'm a little bit confused about the `TExtra` and `TInput`. Is `TExtra` an attempt to simplify the list of the generics by moving more stuff into a single generic? If yes - then this one here should be smth like:
```suggestion
  ) => Model<TContext, TEvent, TAction, TModelCreators, TExtra & { input: TInput }>;
```
or maybe even this:
```suggestion
  ) => Model<TContext, TEvent, TAction, TModelCreators, Omit<TExtra, 'input'> & { input: TInput }>;
```

What should happen here?
```js
createModel({}, {}).withInput({ foo: '' }).withInput({ foo: 42 })
```

I guess the last one should win? Or should they be merged?

As long as we don't allow access to a runtime value somewhere within the `createModel` call having the `TInput` to be defined "later" should be OK-ish I guess.

Reply:
I think the last one should win, meaning your second suggestion should be merged in